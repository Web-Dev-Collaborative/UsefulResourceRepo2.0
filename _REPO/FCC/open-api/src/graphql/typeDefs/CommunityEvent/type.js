import InputTypes from './input';

export default `
type CommunityEvent {
  externalId: String
  title: String!
  description: String!
  owner: User!
  attendees: [User]
  date: String!
  imageUrl: String
}

${InputTypes}

`;
