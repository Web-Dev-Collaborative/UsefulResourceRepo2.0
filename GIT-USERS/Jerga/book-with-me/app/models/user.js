import DS from "ember-data";

export default DS.Model.extend({
  username: DS.attr(),
  email: DS.attr(),
  password: DS.attr(),
  password_confirmation: DS.attr(),
  rentals: DS.hasMany("rental"),
  bookings: DS.hasMany("booking"),
});
