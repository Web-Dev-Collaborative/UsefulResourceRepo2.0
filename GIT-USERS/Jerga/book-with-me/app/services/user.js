import Service from "@ember/service";
import { inject as service } from "@ember/service";

export default Service.extend({
  store: service(),
  session: service("session"),
  currentUser: null,

  initUser() {
    const currentUser = this.get("currentUser");

    if (!currentUser) {
      this.get("session.store")
        .restore()
        .then(({ authenticated }) => {
          const parsedToken = this.parseJwt(authenticated.token);

          return this.createUser(parsedToken);
        });
    }
  },

  createUser(token) {
    let user = {};

    if (token.username && token.email) {
      user = this.get("store").peekRecord("user", token.userId);

      if (!user) {
        user = this.get("store").createRecord("user", {
          id: token.userId,
          username: token.username,
          email: token.email,
        });
      }

      this.set("currentUser", user);
    }

    return user;
  },

  getUser() {
    return this.get("currentUser");
  },

  parseJwt(token) {
    if (token) {
      const base64Url = token.split(".")[1];
      const base64 = base64Url.replace("-", "+").replace("_", "/");

      return JSON.parse(window.atob(base64));
    }

    return {};
  },
});
