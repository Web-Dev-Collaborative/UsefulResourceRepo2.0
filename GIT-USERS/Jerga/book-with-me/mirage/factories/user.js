import { Factory } from "ember-cli-mirage";
import bcrypt from "npm:bcryptjs";

export default Factory.extend({
  id(i) {
    return i;
  },
  username: "Test User",
  email: "test@gmail.com",
  password() {
    const salt = bcrypt.genSaltSync(10);
    return bcrypt.hashSync("testtest", salt);
  },
});
