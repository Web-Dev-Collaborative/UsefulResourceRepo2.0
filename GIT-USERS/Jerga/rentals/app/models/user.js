import DS from 'ember-data';

export default DS.Model.extend({
  name: DS.attr(),
  email: DS.attr(),
  password: DS.attr(),
  password_confirmation: DS.attr(),
  rentals: DS.hasMany('rental', { async: true }),
  bookings: DS.hasMany('bookings')
});
