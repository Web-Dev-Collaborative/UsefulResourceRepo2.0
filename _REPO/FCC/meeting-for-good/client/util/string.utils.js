const nameInitials = (name) => {
  const initials = name.match(/\b\w/g) || [];
  return ((initials.shift() || '') + (initials.pop() || '')).toUpperCase();
};

export default nameInitials;
