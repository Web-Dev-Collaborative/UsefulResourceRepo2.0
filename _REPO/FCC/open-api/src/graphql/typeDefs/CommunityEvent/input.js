// Users by their external id or email address
export default `
input UserInput {
  externalId: String
  email: String
}
`;
