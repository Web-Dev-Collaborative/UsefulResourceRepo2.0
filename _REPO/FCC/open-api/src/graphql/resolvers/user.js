import { createUser, getUser, deleteUser } from '../../dataLayer/mongo/user';

export const userResolvers = {
  Query: {
    getUser
  },
  Mutation: {
    createUser,
    deleteUser
  }
};
