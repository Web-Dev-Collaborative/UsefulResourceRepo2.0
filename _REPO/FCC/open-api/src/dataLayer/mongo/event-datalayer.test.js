/* global expect beforeAll afterAll */
import {
  createCommunityEvent,
  getCommunityEvent,
  getCommunityEvents,
  deleteCommunityEvent
} from './communityEvent';
import { createUser } from './user';
import CommunityEvent from '../model/communityEvent.js';
import { isObject, isEmpty } from 'lodash';
import mongoose from 'mongoose';
import uuid from 'uuid/v4';
import { isDate, isArray } from 'util';

const validContextForBrian = global.mockedContextWithValidTokenForBrian;
const validContextForDennis = global.mockedContextWithValidTokenForDennis;
const validContextForKen = global.mockedContextWithValidTokenForKen;

const event = {
  title: 'epoch',
  description: 'The start of POSIX time',
  date: 'Thu 1 Jan 1970 00:00:00'
};

const eventValidAttendees = {
  ...event,
  attendees: [
    { email: 'dennisritchie@example.com' },
    { email: 'kenthompson@example.com' }
  ]
};

const eventInValidAttendees = {
  ...event,
  attendees: [{ email: 'yeahnah@example.com' }]
};

beforeAll(async function beforeAllTests() {
  await mongoose.connect(global.__MONGO_URI__);
  // Create some users for our tests
  await createUser({}, {}, validContextForBrian);
  await createUser({}, {}, validContextForDennis);
  await createUser({}, {}, validContextForKen);

  // Create an event
  await createCommunityEvent({}, eventValidAttendees, validContextForBrian);
});

afterAll(async function afterAllTests() {
  await mongoose.connection.db.dropDatabase();
  await mongoose.disconnect();
});

describe('createCommunityEvent', () => {
  it('should return an Event object', done => {
    expect.assertions(7);
    createCommunityEvent({}, eventValidAttendees, validContextForBrian)
      .then(result => {
        const {
          externalId,
          title,
          description,
          owner,
          date,
          attendees
        } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()

        // TODO: DRY this please
        const hasKeys =
          !isEmpty(externalId) &&
          !isEmpty(title) &&
          !isEmpty(description) &&
          !isEmpty(owner) &&
          isDate(date) &&
          isArray(attendees);

        const hasOwnerKeys = !isEmpty(owner.name) && !isEmpty(owner.email);
        const hasAttendeeKeys =
          !isEmpty(attendees[0].name) && !isEmpty(attendees[0].email);

        expect(isObject(result)).toBe(true);
        expect(isObject(result.owner)).toBe(true);
        expect(isObject(result.attendees[0])).toBe(true);
        expect(result.attendees).toHaveLength(2);
        expect(hasKeys).toBe(true);
        expect(hasOwnerKeys).toBe(true);
        expect(hasAttendeeKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should throw if an attendee does yet exist', done => {
    expect.assertions(2);
    createCommunityEvent({}, eventInValidAttendees, validContextForBrian).catch(
      err => {
        expect(err).toMatchSnapshot();
        expect(err).toContain('Unable to find attendee');
        done();
        return;
      }
    );
  });

  it('should throw if an imageUrl is not a valid URL', done => {
    expect.assertions();
    createCommunityEvent(
      {},
      {
        ...event,
        imageUrl: 'notaUrl'
      },
      validContextForBrian
    ).catch(err => {
      expect(err).toMatchSnapshot();
      expect(err).toContain('Expected valid URL string');
      done();
      return;
    });
  });

  it('should throw if an isLocked is not a Boolean', done => {
    expect.assertions(2);
    createCommunityEvent(
      {},
      {
        ...event,
        isLocked: 'notaBoolean'
      },
      validContextForBrian
    ).catch(err => {
      expect(err).toMatchSnapshot();
      expect(err).toContain('Expected a Boolean value');
      done();
      return;
    });
  });
});

describe('getCommunityEvent', () => {
  it('should return an Event object for a valid request', done => {
    expect.assertions(7);
    getCommunityEvent({}, { title: 'epoch' }, validContextForBrian)
      .then(result => {
        const {
          externalId,
          title,
          description,
          owner,
          date,
          attendees
        } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(externalId) &&
          !isEmpty(title) &&
          !isEmpty(description) &&
          !isEmpty(owner) &&
          isDate(date) &&
          isArray(attendees);

        const hasOwnerKeys = !isEmpty(owner.name) && !isEmpty(owner.email);
        const hasAttendeeKeys =
          !isEmpty(attendees[0].name) && !isEmpty(attendees[0].email);

        expect(isObject(result)).toBe(true);
        expect(isObject(owner)).toBe(true);
        expect(isObject(attendees[0])).toBe(true);
        expect(attendees).toHaveLength(2);
        expect(hasKeys).toBe(true);
        expect(hasOwnerKeys).toBe(true);
        expect(hasAttendeeKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('title search non existing event should return null', done => {
    expect.assertions(1);
    getCommunityEvent({}, { title: 'Yeah nah' }, validContextForBrian).then(
      data => {
        expect(data).toBe(null);
        done();
      }
    );
  });

  it('externalId search for non existing event should return null', done => {
    expect.assertions(1);
    getCommunityEvent({}, { externalId: uuid() }, validContextForBrian).then(
      data => {
        expect(data).toBe(null);
        done();
      }
    );
  });

  // function validateError(err) {
  //   expect(err).toBeInstanceOf(TypeError);
  //   expect(err.message).toContain('Expected a valid externalId or title');
  // }

  function getCommunityEventPromiseForTestCase(testCase) {
    return getCommunityEvent(
      {},
      { externalId: testCase },
      validContextForBrian
    ).catch(err => {
      expect(err).toBeInstanceOf(TypeError);
      expect(err.message).toContain('Expected a valid externalId or title');
    });
  }

  // TODO: DRY please
  it('should throw if the externalId is not valid', done => {
    expect.assertions(8);
    Promise.all([
      getCommunityEventPromiseForTestCase(1),
      getCommunityEventPromiseForTestCase('abc'),
      getCommunityEventPromiseForTestCase(['yeah nah']),
      getCommunityEventPromiseForTestCase(false)
    ]).then(() => {
      done();
    });
  });

  it('should throw if the title is not valid', done => {
    expect.assertions(6);
    Promise.all([
      getCommunityEvent({}, { title: 1 }, validContextForBrian).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid externalId or title');
      }),
      getCommunityEvent(
        {},
        { title: ['yeah nah'] },
        validContextForBrian
      ).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid externalId or title');
      }),
      getCommunityEvent({}, { title: false }, validContextForBrian).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid externalId or title');
        }
      )
    ]).then(() => {
      done();
    });
  });
});

describe('getCommunityEvents', () => {
  it('should return multiple Event objects for a valid request', done => {
    expect.assertions(8);
    getCommunityEvents({}, { title: 'epoch' }, validContextForBrian)
      .then(result => {
        const {
          externalId,
          title,
          description,
          owner,
          date,
          attendees
        } = result[0];

        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(externalId) &&
          !isEmpty(title) &&
          !isEmpty(description) &&
          !isEmpty(owner) &&
          isDate(date) &&
          isArray(attendees);

        const hasOwnerKeys = !isEmpty(owner.name) && !isEmpty(owner.email);
        const hasAttendeeKeys =
          !isEmpty(attendees[0].name) && !isEmpty(attendees[0].email);

        // TODO: 2 depend on other tests to create events for us,
        // this is a bit brittle..
        expect(result).toHaveLength(3);
        expect(isObject(result)).toBe(true);
        expect(isObject(owner)).toBe(true);
        expect(isObject(attendees[0])).toBe(true);
        expect(attendees).toHaveLength(2);
        expect(hasKeys).toBe(true);
        expect(hasOwnerKeys).toBe(true);
        expect(hasAttendeeKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('title search non existing event should return null', done => {
    expect.assertions(1);
    getCommunityEvents({}, { title: 'Yeah nah' }, validContextForBrian).then(
      data => {
        expect(data).toBe(null);
        done();
      }
    );
  });

  it('externalId search for non existing event should return null', done => {
    expect.assertions(1);
    getCommunityEvents({}, { externalId: uuid() }, validContextForBrian).then(
      data => {
        expect(data).toBe(null);
        done();
      }
    );
  });

  // TODO: DRY please
  it('should throw if the externalId is not valid', done => {
    expect.assertions(8);
    Promise.all([
      getCommunityEvents({}, { externalId: 1 }, validContextForBrian).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid externalId or title');
        }
      ),
      getCommunityEvents({}, { externalId: 'abc' }, validContextForBrian).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid externalId or title');
        }
      ),
      getCommunityEvents(
        {},
        { externalId: ['yeah nah'] },
        validContextForBrian
      ).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid externalId or title');
      }),
      getCommunityEvents({}, { externalId: false }, validContextForBrian).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid externalId or title');
        }
      )
    ]).then(() => {
      done();
    });
  });

  it('should throw if the title is not valid', done => {
    expect.assertions(6);
    Promise.all([
      getCommunityEvents({}, { title: 1 }, validContextForBrian).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid externalId or title');
      }),
      getCommunityEvents(
        {},
        { title: ['yeah nah'] },
        validContextForBrian
      ).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid externalId or title');
      }),
      getCommunityEvents({}, { title: false }, validContextForBrian).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid externalId or title');
        }
      )
    ]).then(() => {
      done();
    });
  });
});

describe('deleteCommunityEvent', () => {
  it('should delete an existing event', async done => {
    const event = {
      title: 'deleteCommunityEvent event',
      description: 'A boring test event',
      date: 'Thu 1 Jan 1970 00:00:00'
    };

    const createdEvent = await createCommunityEvent(
      {},
      event,
      validContextForBrian
    );
    const deletedEvent = await deleteCommunityEvent(
      {},
      { externalId: createdEvent.externalId },
      validContextForBrian
    );

    expect(createdEvent.externalId).toMatch(deletedEvent.externalId);

    const foundEvent = await CommunityEvent.findOne({
      externalId: deletedEvent.externalId
    }).exec();
    expect(foundEvent).toBe(null);
    done();
  });

  it('should return with an error for a non existing event', async done => {
    try {
      await deleteCommunityEvent(
        {},
        { externalId: uuid() },
        validContextForBrian
      );
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toContain('Event not found');
      expect(err).toMatchSnapshot();
    }
    done();
  });

  it('should refuse deletion of events owned by other users', async done => {
    const e = await CommunityEvent.findOne({ title: 'epoch' }).exec();
    try {
      await deleteCommunityEvent(
        {},
        { externalId: e.externalId },
        validContextForKen
      );
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toContain('Only allowed to delete events you own');
      expect(err).toMatchSnapshot();
    }
    done();
  });
});
