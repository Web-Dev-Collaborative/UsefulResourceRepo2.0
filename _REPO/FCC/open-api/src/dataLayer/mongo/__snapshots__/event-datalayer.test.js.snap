// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCommunityEvent should throw if an attendee does yet exist 1`] = `"Unable to find attendee: {\\"email\\":\\"yeahnah@example.com\\"}"`;

exports[`createCommunityEvent should throw if an imageUrl is not a valid URL 1`] = `"Expected valid URL string, got \\"notaUrl\\""`;

exports[`createCommunityEvent should throw if an isLocked is not a Boolean 1`] = `"Expected a Boolean value, got \\"notaBoolean\\""`;

exports[`deleteCommunityEvent should return with an error for a non existing event 1`] = `[Error: Event not found]`;
